package vanet_pb;
option java_package = "edu.psu.cse.vadroid";

enum Resource
{
	MODEL = 0;
	WEIGHTS = 1;
	MEAN = 2;
	SYNSET = 3;
}

enum ProcessMode
{
	MOBILE = 1;
	SERVER = 2;
	BOTH = 3;
}

message VideoInfo
{
	required string path = 1;
	optional string name = 2;
	optional string timestamp = 3;
	optional uint64 size = 4;
	optional uint64 duration = 5;
	optional uint32 bitrate = 6;
	optional string mime = 7;
	optional double loc_lat = 8;
	optional double loc_long = 9;
	optional uint32 width = 10;
	optional uint32 height = 11;
	optional uint32 rotation = 12;
	repeated float class_probs = 13;
	optional uint32 frames_processed = 14;
	optional uint32 frames_total = 15;
	optional ProcessMode process_mode = 16;
	optional uint32 extraction_time = 17;
	optional uint32 classification_time = 18;
	optional uint32 est_process_time = 19;
	repeated uint32 tags = 20;
}

message PartialInfo
{
	required string path = 1;
	optional bool final = 2;
	optional uint32 frame_size = 3;
	optional uint32 batch_size = 4;
	optional ProcessMode process_mode = 5;
	repeated float class_probs = 6;
	repeated uint32 tags = 7;
}

message ServerMessage
{
	enum Type
	{
		CONNECT = 0;
		INIT = 1;
		RES = 2;
		QUERY_ALL = 3;
		QUERY_TAG = 4;
		QUERY_TIME = 5;
		QUERY_LOCATION = 6;
		PROCESS_DIRECTIVE = 7;
		VIDEO_REQUEST = 8;
		PING = 9;
		DISCONNECT = 10;
	}

	required Type type = 1;
	repeated string path = 2;
	repeated uint32 tags = 3;
	optional string start_time = 4;
	optional string end_time = 5;
	optional double loc_lat = 6;
	optional double loc_long = 7;
	repeated ProcessMode process_mode = 8;
	repeated Resource resources = 9;
	repeated uint64 size = 10;
	optional bool ack = 11;
	optional string ip = 12;
	optional uint32 top_k = 13;
}

message ClientMessage
{
	enum Type
	{
		CONNECT = 0;
		INIT = 1;
		VIDEO = 2;
		VIDEO_INFO = 3;
		PARTIAL_INFO = 4;
		FRAME = 5;
		PING = 6;
		UPLINK_TEST = 7;
		DISCONNECT = 8;
	}

	required uint64 mac = 1;
	required Type type = 2;
	optional int32 size = 3;
	repeated VideoInfo video_info = 4;
	optional PartialInfo partial_info = 5;
	repeated Resource resources = 6;
	optional bool terminate = 7;
	optional bool ack = 8;
}